urlfb310a588a_alokhan_LinkStateNetworkSimulation_tgz-pJ8-reso_examples_static_routing_DemoJ8/src/linkStateRouting/LinkStateRoutingProtocol.java

```diff
@@ -0,0 +1,295 @@
+                    if (iface.isActive()) {
+                        neighborList.put(ls.getKey(), new LinkState(ls.getValue().routerId, (Integer) iface.getAttribute("metric"), ls.getValue().routerInterface));
+                    } else {
+                        keyToRemove = ls.getKey();
+                    }
+                }
+            }
+            if (keyToRemove != null) {
+                neighborList.remove(keyToRemove);
+            }
+        }
+    }
+
+    /**
+     * Send HelloMessage on all our interfaces.
+     *
+     * @throws Exception
+     */
+    public void SendHelloToAllInterfaces() throws Exception {
+        for (IPInterfaceAdapter iface : ip.getInterfaces()) {
+            if (iface instanceof IPLoopbackAdapter) {
+                continue;
+            }
+            HelloMessage hello = new HelloMessage(getRouterID(), neighborList.keySet());
+            iface.send(new Datagram(iface.getAddress(), IPAddress.BROADCAST, IP_PROTO_LS, 1, hello), null);
+        }
+    }
+
+    /**
+     * Send LinkStateMessage on all our interfaces
+     *
+     * @throws Exception
+     */
+    public void SendLSPToAllInterfaces() throws Exception {
+        for (IPInterfaceAdapter iface : ip.getInterfaces()) {
+            if (iface instanceof IPLoopbackAdapter) {
+                continue;
+            }
+            LinkStateMessage LSM = new LinkStateMessage(getRouterID());
+            for (LinkState ls : neighborList.values()) {
+                LSM.addLS(ls);
+            }
+            LSDB.put(getRouterID(), LSM);
+            iface.send(new Datagram(iface.getAddress(), IPAddress.BROADCAST, IP_PROTO_LS, 1, LSM), null);
+        }
+    }
+
+    /**
+     * Send LinkStateMessage on all our interfaces but the one we got it from.
+     *
+     * @param src
+     * @param msg
+     * @throws Exception
+     */
+    private void SendToAllButSender(IPInterfaceAdapter src, LinkStateMessage msg) throws Exception {
+        for (IPInterfaceAdapter iface : ip.getInterfaces()) {
+            if (iface.equals(src) || iface instanceof IPLoopbackAdapter) {
+                continue;
+            }
+            Datagram newDatagram = new Datagram(iface.getAddress(), IPAddress.BROADCAST, IP_PROTO_LS, 1, msg);
+            iface.send(newDatagram, null);
+        }
+    }
+
+    private IPAddress getRouterID() {
+        IPAddress routerID = null;
+        for (IPInterfaceAdapter iface : ip.getInterfaces()) {
+            IPAddress addr = iface.getAddress();
+            if (routerID == null) {
+                routerID = addr;
+            } else if (routerID.compareTo(addr) < 0) {
+                routerID = addr;
+            }
+        }
+        return routerID;
+    }
+
+    /**
+     * Calculate the shortest path from a router to all the other routers.
+     *
+     * @param LSDB
+     * @throws Exception
+     */
+    private void Compute(Map<IPAddress, LinkStateMessage> LSDB) throws Exception {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 259-273, lines 276-292 into separate methods.